src/mtr.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+  Adplug - Replayer for many OPL2/OPL3 audio file formats.
+  Copyright (C) 1999 - 2003 Simon Peter, <dn.tlp@gmx.net>, et al.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+  mtr.cpp - MTR loader by Dmitry Smagin <dmitry.s.smagin@gmail.com>
+*/
+
+#include <stdio.h>
+#include <cstring>
+#include "mtr.h"
+#include "debug.h"
+
+/* -------- Public Methods -------------------------------- */
+
+CPlayer *CmtrLoader::factory(Copl *newopl) {
+    return new CmtrLoader(newopl);
+}
+
+bool CmtrLoader::load(const std::string &filename, const CFileProvider &fp) {
+    binistream *f = fp.open(filename);
+    if (!f)
+        return false;
+
+    unsigned int i, j, k, t = 0;
+    ninstruments = 0;
+
+    char header[51] = {0};
+    char mtitle[21] = {0};
+    int nvoices, ndigvoices, npatterns, orderlen,
+        timervalue = 0x428F, restart, flength;
+
+    f->readString(header, 50);
+
+    if (!strncmp(header, "MTRAC ", 6)) {
+        version = 1;
+        sscanf(header + 26, "%02x %02x %02x %02x %08x",
```

Comment:
Need to error-out if sscan() returns another value than 5

(Which will be the case for .MTR (v1.x) files)

Reply:
v1 files have "MTRACK\x20" at the beginning, v2 have "MTRACK\x20NC\x20".
sscanf shouldn't fail once the version is recognized.
AKMTRK2.DOC is incorrect describing this. There are lots of errors there, too much to enumerate them all, easier to create another 'fixed' variant of a doc.

Reply:
Fixed.